### PR TITLE
Hclang 53 ebnf explanation

### PR DESCRIPTION
**Pull request by Lambcreek**
### Description
Added stuff about EBNF to report.

### Changelog
 - Added EBNF to appendix
   - EBNF has been changed (in report) because we noticed some errors the other day - Casper's branch has changes to .txt file
 - Theory section describing what CFG and EBNF are
 - EBNF has been used to complement some of the text in LanguageSpecification.tex

### Issues solved 
 - HCLANG-53

## FOR REVIEWER [Do not remove]
A small checklist of things to note when you are doing a review. Make sure these things apply before going into master.
 - Did the PR actually fix the issue
 - Comment anything that doesn't make sense or anything you want clarified. Many comments are always ok
 - Make sure this is ready for exams if this is going into master
 - It's always a good idea having another reviewer look at a PR as well, if it's major. get a buddy.

### Regarding filenames
- all .tex files use pascal case
- all other files [except in /code/] use snake case

### Regarding Report [if applicable]
- Does each line have a linechange in the sourcecode?
- Does sentences make sense?
- Did you do spellchecking?
- Citations?

### CODE REVIEW

File: report/3.Theory/ContextFreeGrammars.tex [hunk 1]
```diff
@@ -0,0 +1,130 @@
+\section{Context-free Grammars}
+\label{CFGdescription}
+Context-free grammars (CFGs) are a method for describing syntactical composition of sentences in a language.
+It does not describe the meaning (semantics) of the sentences, hence the name, "context-free".
+CFGs split sentences into "terminals" and "non-terminals" (also known as variables).
```
Comment (C0DK) [inline report/3.Theory/ContextFreeGrammars.tex]: syntactic variables

File: report/3.Theory/ContextFreeGrammars.tex [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+	\item one terminal, "word",
+	\item the empty string (denoted by a lambda, $\lambda$),
+	\item and a dollar-sign, \$, marking the end of the string.
+\end{itemize}
+
+The rules of the grammar are as follows:
+
+Sentence consists of one StringOfWords. 
+A StringOfWords consists of the string "word" followed by another StringOfWords, or an empty string.
+As StringOfWords appears on the right-hand side of the arrow in its own definition, it is recursive.
```
Comment (C0DK) [inline report/3.Theory/ContextFreeGrammars.tex]: i would change it to "therefore it is recursive"

File: report/3.Theory/LanguageSpecification.tex [hunk 3]
```diff
@@ -76,6 +85,27 @@ \section{Language Specification}
 ["john", "frank"] where {value is "frank"} print
 \end{lstlisting}
 
+\textbf{Grammar for function calls}\\
+There are three different ways to call a function depending on how many parameters must be passed with the function:
+\begin{itemize}
+	\item If no parameters must be passed, then invoking the function identifier is a call in and of itself.
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: "If the function has no parameters, then invokation of the function is done with the identifier alone. an example could be "printHelloWorld"

File: report/3.Theory/LanguageSpecification.tex [hunk 4]
```diff
@@ -76,6 +85,27 @@ \section{Language Specification}
...
+	\item If one parameter must be passed, then the parameter must appear on the left-hand side of the function identifier.
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: "argument" er input - parameter er inde i metoden så der skal stå 

"If one parameter must be passed, then the argument must appear on the left-hand side of the function identifier."

File: report/3.Theory/LanguageSpecification.tex [hunk 5]
```diff
@@ -76,6 +85,27 @@ \section{Language Specification}
...
+	\item If more than one parameter must be passed, the first parameter must appear on the left-hand side of the function identifier, and the rest must appear, in order, on the right-hand side of the function.
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: also "argument" here

File: report/3.Theory/LanguageSpecification.tex [hunk 6]
```diff
@@ -76,6 +85,27 @@ \section{Language Specification}
 ["john", "frank"] where {value is "frank"} print
 \end{lstlisting}
 
+\textbf{Grammar for function calls}\\
+There are three different ways to call a function depending on how many parameters must be passed with the function:
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: maybe explain infix, so something like "first argument is on the left side, while the remaining is on the right".

whole sentence could be restructured to something along the lines of

"Function calls are done with infix notation, meaning that if a given method is defined with parameters, the first argument has to be in front of the function identifier, while all remaining arguments are postfixed.

As explained by the following examples:
- printHelloWorld
- "hello world :)" print
- "hello" printWords ["world",":)"]

File: report/3.Theory/LanguageSpecification.tex [hunk 7]
```diff
@@ -76,6 +85,27 @@ \section{Language Specification}
...
+As it is not allowed to have parameters on the right-hand side of the identifier without also having one on the left-hand side, this rule cannot accurately be described with only one production.
+
+The first production shows how the function is called if it has no parameters.
+The second production shows that if the function has parameters, there must always be exactly one parameter on the left-hand side of the identifier, and optionally a sequence of parameters on the right-hand side.
+
+The non-terminal \texttt{<FunCallParam>} refers to any legal parameter for function calls, including identifiers, literals, lambda expressions, and other function calls.
+
+The non-terminal \texttt{<FunCallParams>} refers to a sequence of at least one \texttt{<FunCallParam>}, optionally enclosed in parentheses.
+
 \textbf{Loops}\\
```
Comment (C0DK) [inline report/3.Theory/LanguageSpecification.tex]: explain that loops are in fact a function just as everything else

Comment (Abildgaard96) [reply report/3.Theory/LanguageSpecification.tex]: kinda explained in line 111, but could be more explicit

Comment (C0DK) [reply report/3.Theory/LanguageSpecification.tex]: fair

**Pull request merged**